homeassistant/components/anomaly/binary_sensor.py

```diff
@@ -0,0 +1,326 @@
+"""A sensor that detects anomalies in other components."""
+from collections import deque
+import logging
+from typing import Optional
+
+import voluptuous as vol
+
+from homeassistant.components.binary_sensor import (
+    DEVICE_CLASSES_SCHEMA,
+    ENTITY_ID_FORMAT,
+    PLATFORM_SCHEMA,
+    BinarySensorEntity,
+)
+from homeassistant.const import (
+    ATTR_ENTITY_ID,
+    ATTR_FRIENDLY_NAME,
+    CONF_DEVICE_CLASS,
+    CONF_ENTITY_ID,
+    CONF_FRIENDLY_NAME,
+    CONF_SENSORS,
+    STATE_UNAVAILABLE,
+    STATE_UNKNOWN,
+)
+from homeassistant.core import State, callback
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import generate_entity_id
+from homeassistant.helpers.event import async_track_state_change_event
+from homeassistant.helpers.reload import setup_reload_service
+from homeassistant.util import utcnow
+
+from . import DOMAIN, PLATFORMS
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_ATTRIBUTE = "attribute"
+ATTR_INVERT = "invert"
+ATTR_REQUIRE_BOTH = "require_both"
+ATTR_POSITIVE_ONLY = "positive_only"
+ATTR_NEGATIVE_ONLY = "negative_only"
+ATTR_MIN_CHANGE_AMOUNT = "min_change_amount"
+ATTR_MIN_CHANGE_PERCENT = "min_change_percent"
+ATTR_SAMPLE_DURATION = "sample_duration"
+ATTR_MAX_SAMPLES = "max_samples"
+ATTR_TRAILING_SAMPLE_DURATION = "trailing_sample_duration"
+ATTR_MAX_TRAILING_SAMPLES = "max_trailing_samples"
+ATTR_SAMPLE_COUNT = "sample_count"
+ATTR_TRAILING_SAMPLE_COUNT = "trailing_sample_count"
+ATTR_SAMPLE_AVG = "sample_average"
+ATTR_TRAILING_AVG = "trailing_sample_average"
+ATTR_CHANGE_AMOUNT = "change_amount"
+ATTR_CHANGE_PERCENT = "change_percent"
+
+CONF_ATTRIBUTE = "attribute"
+CONF_INVERT = "invert"
+CONF_POSITIVE_ONLY = "positive_only"
+CONF_NEGATIVE_ONLY = "negative_only"
+CONF_REQUIRE_BOTH = "require_both"
+CONF_MAX_SAMPLES = "max_samples"
+CONF_SAMPLE_DURATION = "sample_duration"
+CONF_MAX_TRAILING_SAMPLES = "max_trailing_samples"
+CONF_TRAILING_SAMPLE_DURATION = "trailing_sample_duration"
+CONF_MIN_CHANGE_AMOUNT = "min_change_amount"
+CONF_MIN_CHANGE_PERCENT = "min_change_percent"
+
+SENSOR_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_ENTITY_ID): cv.entity_id,
+        vol.Optional(CONF_ATTRIBUTE): cv.string,
+        vol.Optional(CONF_DEVICE_CLASS): DEVICE_CLASSES_SCHEMA,
+        vol.Optional(CONF_FRIENDLY_NAME): cv.string,
```

Comment:
Can we use the configured entity_id as a unique_id for this sensor and let the user change name via the GUI instead?

Reply:
this entity_id is the entity we are monitoring. so this would be binary_sensor.(same name as the sensor we are monitoring)?

Reply:
device_id that was being passed to the class init was only being used in the generation of the entity_id of the instance of this sensor. the entity_id being passed in, is the entity_id of the device we are monitoring. I will make that more explicit and move the entity_id creation to the setup.

Reply:
and again, I was using the trend sensor as a template for this sensor. If there is a better one for me to use, please let me know. I haven't found any built in sensors that were added recently and use these newer design patterns.

Reply:
I'm asking if we can find and set a unique_id for this entity. Then we can remove the friendly name option.

Reply:
So, I am finally able to swing back around to this. I get where you are going on this, but I don't see a good way to set a unique ID when creating something from a YAML file. The user will always be able to change what is in there, thus causing them to lose the config. Do you have an integration that does this that I can use as a reference point?

Reply:
@MartinHjelmare should I switch to the config flow instead of yaml config?

Reply:
That's not necessary but we should put the config under the integration domain key instead of under the binary sensor domain.

Please move the config schema to the integration init module and set up the binary sensor via our discovery helper `async_load_platform`.

Reply:
And instead of configuring friendly name we should allow configuring a unique id. Then the user can change name and entity id of this entity in the GUI.

Reply:
I mirrored the template sensor config. Allowing one to set the friendly name in the config, but also allow overriding via GUI with a unique ID.